Kernel/Devices/KeyboardDevice.cpp

```diff
@@ -237,15 +237,32 @@ static KeyCode shifted_key_map[0x100] = {
 
 void KeyboardDevice::key_state_changed(u8 raw, bool pressed)
 {
+    bool just_pressed = false, just_released = false;
+    char character = (m_modifiers & Mod_Shift) ? shift_map[raw] : map[raw];
+    KeyCode key = (m_modifiers & Mod_Shift) ? shifted_key_map[raw] : unshifted_key_map[raw];
+
+    if (m_last_state[key] != pressed)
+    {
+        if (pressed)
+            just_pressed = true;
+        else
+            // FIXME: Very unreliable because release IRQ sometimes gets lost somewhere.
```

Comment:
Why are the IRQ's getting lost?
Also, why do you need the `just_pressed` and `just_released` states?
Why can't this be implemented in the same way that Alt, Ctrl and Shift are implemented?

Reply:
We can't use all characters from `shift_map` since caps lock only transforms letters and nothing else but we could just use the upper case letter part of it, good point. 
We don't necessarily have to plumb it to LibGUI but that was just a possible improvement in future. I can't think of a usecase right now but knowing its state might be useful for some userland apps.
I'm not exactly sure why IRQs are getting lost but I just wasn't receiving the up/release IRQ after holding down a letter for more than a second or so. This needs more investigation and I could look into it further but it doesn't impact the Caps Lock feature.
I implemented the `just_pressed` and `just_released` to know exactly when a key is pressed and released, not just whether it's held down or not but I just tested this again and they're not actually necessary for this Caps Lock implementation but the game developer in me still thinks they might be useful at some point in some userland app. I've only just realized Caps Lock key unlike other (letter) keys doesn't generate IRQs while it's held down so we could just use that instead. I could also remove these flags if you don't want them now.

Reply:
Ah, fair point about `shift_map`, duh :)

More generally, I don't think we should be adding any flags if we're not going to be using them.

I know it's fun to hack on random stuff, but please only add things that you want to use. If you don't have a feature or fix in mind when adding something, then it's usually better to leave it alone.

Reply:
I see, makes sense. I will clean this up now. Do I just close this PR and just make a new clean one or nuke commits in this one and do it here?